### ISSUE TITLE
F841 does not have branch analysis

### ISSUE THREAD

**Issue by sk-** [astral-sh/ruff#16752]
### Summary

F841 does not report variables that are unused in certain branches. Below is a short example ([playground](https://play.ruff.rs/0e1d38e6-17b0-4355-a37a-089d454de311)) showing the issue

```python
def func(x):
    if x:
        io = ""
    else:
        io = "foo"
        print(io)   
```

This issue was [originally reported](https://github.com/PyCQA/pyflakes/issues/758) to pyflakes and closed as duplicate of https://github.com/PyCQA/pyflakes/issues/715.

They mention is unlikely they will implement branch analysis, but ruff already has (I guess) the machinery needed to support this use case.

### Version

v0.11.0 (playground)

**Comment by MichaReiser** [comment]
> They mention is unlikely they will implement branch analysis, but ruff already has (I guess) the machinery needed to support this use case.

Actually, we don't but we're working on it. But it's going to take a while to get ready for prime time ;)
